cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -496,28 +498,42 @@ private Location createTempLocationDirectory() throws IOException {
    */
   @SuppressWarnings("unchecked")
   private void beforeSubmit(final Job job) throws TransactionFailureException {
+
+    // AbstractMapReduce implements final initialize(context) and requires subclass to
+    // implement initialize(), whereas programs that directly implement MapReduce have
+    // the option to override initialize(context) (if they implement ProgramLifeCycle)
+    final TransactionControl txControl = mapReduce instanceof AbstractMapReduce
+      ? Transactions.getTransactionControl(TransactionControl.IMPLICIT, AbstractMapReduce.class,
+                                           mapReduce, "initialize")
+      : mapReduce instanceof ProgramLifecycle
+      ? Transactions.getTransactionControl(TransactionControl.IMPLICIT, MapReduce.class,
+                                           mapReduce, "initialize", MapReduceContext.class)
+      : TransactionControl.IMPLICIT;
+
+    if (TransactionControl.EXPLICIT == txControl) {
+      ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(job.getConfiguration().getClassLoader());
+      try {
+        doInitialize();
+      } catch (Exception e) {
+        throw Throwables.propagate(e);
```

Comment:
good point. It does not save us the try { } block, though, because we need the finally clause.